app/src/main/kotlin/com/wire/android/ui/home/conversations/DownloadedAssetDialog.kt

```diff
@@ -0,0 +1,100 @@
+            rememberLauncherForActivityResult(ActivityResultContracts.RequestPermission()) { isGranted ->
+                if (isGranted) {
+                    onDownloadAsset()
+                } else {
+                    // TODO: Implement denied permission rationale
+                }
+            })
+
+        WireDialog(
+            title = dialogState.assetName ?: stringResource(R.string.asset_download_dialog_default_title),
+            text = stringResource(R.string.asset_download_dialog_text),
+            buttonsHorizontalAlignment = false,
+            onDismiss = { conversationViewModel.hideOnAssetDownloadedDialog() },
+            optionButton2Properties = WireDialogButtonProperties(
+                text = stringResource(R.string.asset_download_dialog_open_text),
+                type = WireDialogButtonType.Primary,
+                onClick = {
+                    openAssetFile(dialogState.assetName, assetData, context) { conversationViewModel.onOpenFileError() }
+                    conversationViewModel.hideOnAssetDownloadedDialog()
+                }
+            ),
+            optionButton1Properties = WireDialogButtonProperties(
+                text = stringResource(R.string.asset_download_dialog_save_text),
+                type = WireDialogButtonType.Primary,
+                onClick = { requestWriteToExternalStorageRequestFlow.launch() }
+            ),
+            dismissButtonProperties = WireDialogButtonProperties(
+                text = stringResource(R.string.label_cancel),
+                onClick = { conversationViewModel.hideOnAssetDownloadedDialog() }
+            ),
+        )
+    }
+}
+
+fun saveFileToDownloadsFolder(assetName: String?, assetData: ByteArray, context: Context) {
+    val file = File(context.getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS), "${System.currentTimeMillis()}_$assetName")
+    file.setWritable(true)
+    file.writeBytes(assetData)
+    context.saveFileDataToDownloadsFolder(file, assetData.size)
+}
+
+private fun openAssetFile(assetName: String?, assetData: ByteArray, context: Context, onError: () -> Unit) {
+    val fileName = "$assetName"
+    val assetUri = copyDataToTempFile(context, fileName, assetData)
+
+    // Set intent and launch
+    val intent = Intent()
+    intent.action = Intent.ACTION_VIEW
+    // These flags allow the external app to access the temporal uri
+    intent.flags = if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.LOLLIPOP) {
+        Intent.FLAG_GRANT_WRITE_URI_PERMISSION or Intent.FLAG_GRANT_READ_URI_PERMISSION
+    } else {
+        Intent.FLAG_GRANT_READ_URI_PERMISSION
+    }
+    intent.setDataAndType(assetUri, assetUri.getMimeType(context))
+
+    try {
+        context.startActivity(intent)
+    } catch (noActivityFoundException: ActivityNotFoundException) {
+        appLogger.e("Couldn't find a proper app to process the asset")
+        onError()
```

Comment:
aren't we suppose to delete the tempFile or something after we cannot do anything with it ?

Reply:
According to docs it is not really needed as the GC will dispose those files in the cache directory:
_Apps are strongly encouraged to keep their usage of cache space below the quota returned by StorageManager.getCacheQuotaBytes(java.util.UUID). If your app goes above this quota, your cached files will be some of the first to be deleted when additional disk space is needed._

Reply:
If we go above the qouta :P but do we want to go above the qouta ? wouldt it be better to remove it manually since we do not need it anyway ?